render-wasm/src/render.rs

```diff
@@ -241,65 +281,69 @@ impl RenderState {
         }
 
-        self.apply_drawing_to_final_canvas();
+        self.apply_drawing_to_render_canvas();
     }
 
-    pub fn zoom(&mut self, tree: &HashMap<Uuid, Shape>) -> Result<(), String> {
-        if let Some(cached_surface_image) = self.cached_surface_image.as_mut() {
-            let is_dirty = cached_surface_image.is_dirty_for_zooming(&self.viewbox);
-            if is_dirty {
-                self.render_all(tree, true);
-            } else {
-                self.render_all_from_cache()?;
+    pub fn start_render_loop(&mut self, tree: &mut HashMap<Uuid, Shape>, timestamp: i32) {
+        if self.render_in_progress {
+            if let Some(frame_id) = self.render_request_id {
+                self.cancel_animation_frame(frame_id);
             }
         }
-
-        Ok(())
-    }
-
-    pub fn pan(&mut self, tree: &HashMap<Uuid, Shape>) -> Result<(), String> {
-        if let Some(cached_surface_image) = self.cached_surface_image.as_mut() {
-            let is_dirty = cached_surface_image.is_dirty_for_panning(&self.viewbox);
-            if is_dirty {
-                self.render_all(tree, true);
-            } else {
-                self.render_all_from_cache()?;
-            }
-        }
-
-        Ok(())
-    }
-
-    pub fn render_all(&mut self, tree: &HashMap<Uuid, Shape>, generate_cached_surface_image: bool) {
         self.reset_canvas();
         self.scale(
             self.viewbox.zoom * self.options.dpr(),
             self.viewbox.zoom * self.options.dpr(),
         );
         self.translate(self.viewbox.pan_x, self.viewbox.pan_y);
+        self.pending_stack = vec![(Uuid::nil(), false)];
+        self.render_in_progress = true;
+        self.process_animation_frame(tree, timestamp);
+    }
 
-        // Reset shape tree
-        let is_complete = self.render_shape_tree(&Uuid::nil(), tree);
-        if generate_cached_surface_image || self.cached_surface_image.is_none() {
-            self.cached_surface_image = Some(CachedSurfaceImage {
-                image: self.final_surface.image_snapshot(),
-                viewbox: self.viewbox,
-                has_all_shapes: is_complete,
-            });
+    pub fn request_animation_frame(&mut self) -> i32 {
+        let script =
+            std::ffi::CString::new("requestAnimationFrame(_process_animation_frame)").unwrap();
+        unsafe { emscripten_run_script_int(script.as_ptr()) }
+    }
+
+    pub fn cancel_animation_frame(&mut self, frame_id: i32) {
+        let cancel_script = format!("cancelAnimationFrame({})", frame_id);
+        let c_cancel_script = std::ffi::CString::new(cancel_script).unwrap();
+        unsafe {
+            emscripten_run_script(c_cancel_script.as_ptr());
         }
+    }
 
-        if self.options.is_debug_visible() {
-            self.render_debug();
+    pub fn process_animation_frame(&mut self, tree: &mut HashMap<Uuid, Shape>, timestamp: i32) {
```

Comment:
Let's make this function return a `Result` instead of handling `.expect`s here

Reply:
Updated!